shared/domains/chelonia/GIMessage.js

```diff
@@ -50,61 +62,143 @@ export type GIOpKeyUpdate = {
-export type GIOp = [GIOpType, GIOpValue] | [GIOpType, GIOpValue, GIOpValue]
+export type GIOpRaw = [GIOpType, SignedData<GIOpValue>]
+export type GIOp = [GIOpType, GIOpValue]
+
+export type GIMsgDirection = 'incoming' | 'outgoing'
+type GIMsgParams = { direction: GIMsgDirection, mapping: Object; head: Object; signedMessageData: SignedData<GIOpValue> }
 
-type GIMsgParams = { mapping: Object; head: Object; message: GIOpValue; decryptedValue?: GIOpValue; signature: string; signedPayload: string }
+// Takes a raw message and processes it so that EncryptedData and SignedData
+// attributes are defined
+const decryptedAndVerifiedDeserializedMessage = (head: Object, headJSON: string, contractID: string, parsedMessage: GIOpValue, additionalKeys?: Object, state: Object): GIOpValue => {
+  const op = head.op
+  const height = head.height
 
-const decryptedDeserializedMessage = (op: string, height: number, parsedMessage: Object, contractID: string, additionalKeys?: Object, state: Object) => {
-  const message = op === GIMessage.OP_ACTION_ENCRYPTED ? encryptedIncomingData(contractID, state, parsedMessage, height, additionalKeys) : parsedMessage
+  const message: GIOpValue = op === GIMessage.OP_ACTION_ENCRYPTED
+    // $FlowFixMe
+    ? encryptedIncomingData<GIOpActionUnencrypted>(contractID, state, (parsedMessage: any), height, additionalKeys, headJSON, undefined)
+    : parsedMessage
+
+  // If the operation is GIMessage.OP_KEY_ADD or GIMessage.OP_KEY_UPDATE,
+  // extract encrypted data from key.meta?.private?.content
   if ([GIMessage.OP_KEY_ADD, GIMessage.OP_KEY_UPDATE].includes(op)) {
-    ((message: any): any[]).forEach((key) => {
-      // TODO: When storing the message, ensure only the raw encrypted data get stored. This goes for all uses of encryptedIncomingData
-      if (key.meta?.private?.content) {
-        key.meta.private.content = encryptedIncomingData(contractID, state, key.meta.private.content, height, additionalKeys, (value) => {
+    return ((message: any): any[]).map((key) => {
+      return maybeEncryptedIncomingData(contractID, state, key, height, additionalKeys, headJSON, (key, eKeyId) => {
+        if (key.meta?.private?.content) {
+          key.meta.private.content = encryptedIncomingData(contractID, state, key.meta.private.content, height, additionalKeys, headJSON, (value) => {
+          // Validator function to verify the key matches its expected ID
+            const computedKeyId = keyId(value)
+            if (computedKeyId !== key.id) {
+              throw new Error(`Key ID mismatch. Expected to decrypt key ID ${key.id} but got ${computedKeyId}`)
+            }
+          })
+        }
+        if (key.meta?.keyRequest?.contractID) {
+          try {
+            key.meta.keyRequest.contractID = maybeEncryptedIncomingData(contractID, state, key.meta.keyRequest.contractID, height, additionalKeys, headJSON)?.valueOf()
+          } catch {
+            delete key.meta.keyRequest.contractID
```

Comment:
Silent error? If an error shouldn't be logged, a comment should always explain why not.

Otherwise, errors should always be logged.